### PR TITLE
move SSE-C TLS enforcement into generic handler

### PR DESCRIPTION
**Pull request by aead**
## Description
This commit moves the check that SSE-C requests
must be made over TLS into a generic HTTP handler.

Since the HTTP server uses custom TCP connection handling
it is not possible to use `http.Request.TLS` to check
for TLS connections. So using `globalIsSSL` is the only
option to detect whether the request is made over TLS.
By extracting this check into a separate handler it's possible
to refactor other parts of the SSE handling code further.

## Motivation and Context
SSE-C, refactoring

## Regression
no

## How Has This Been Tested?
manually, mint 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#17390f79-9a31-4277-83ad-1f69981507c9) |
<a name="17390f79-9a31-4277-83ad-1f69981507c9"></a>
### 6639-9bd7335/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30926
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 967ea3126423:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 1 minutes and 20 seconds
(2/13) Running aws-sdk-go tests ... done in 3 seconds
(3/13) Running aws-sdk-java tests ... done in 10 seconds
(4/13) Running aws-sdk-php tests ... done in 57 seconds
(5/13) Running aws-sdk-ruby tests ... done in 37 seconds
(6/13) Running mc tests ... done in 1 minutes and 7 seconds
(7/13) Running minio-dotnet tests ... FAILED in 4 minutes and 44 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

File: cmd/generic-handlers.go [hunk 1]
```diff
@@ -773,3 +774,17 @@ func (h criticalErrorHandler) ServeHTTP(w http.ResponseWriter, r *http.Request)
 	}()
 	h.handler.ServeHTTP(w, r)
 }
+
+func setTLSHandler(h http.Handler) http.Handler { return tlsHandler{h} }
+
+// tlsHandler enforces certain rules for requests which are made / must be made over TLS.
+type tlsHandler struct{ handler http.Handler }
```
Comment (harshavardhana) [inline cmd/generic-handlers.go]: can we call this sseTLSHandler?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6639-0c037be/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6639-0c037be/gopath/src/github.com/minio/minio
++ cd 6639-0c037be/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6639-0c037be/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6639/head:pr6639
++ cd 6639-0c037be/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6639
++ minikube ip
+ minikube_ip=192.168.39.64
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.64
2018/10/16 19:28:59 wrote public.crt
2018/10/16 19:28:59 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6639-0c037be/gopath
+ make -C 6639-0c037be/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/routers.go:86:2: undefined: setTLSHandler
make: *** [build] Error 2

```

File: cmd/generic-handlers.go (hunk 1 above)
Comment (aead) [reply cmd/generic-handlers.go]: Sure, done. Also added unit test.

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**